File: src/gestorAplicacion/economia/Prestamo.java
package gestorAplicacion.economia;

import administrador.Utils;
import gestorAplicacion.usuario.Cuenta;
import gestorAplicacion.usuario.Usuario;

import java.io.Serializable;
import java.time.LocalDate;

public abstract class Prestamo implements Serializable, Abonable<double[]> {

    /**
     *
     */
    private static final long serialVersionUID = -5514442359859569278L;

    private Usuario usuario;
    protected double valorInicial;
    protected double valorPagado;
    protected int tiempo;            //meses diferidos
    private final LocalDate fechaInicio;
    private LocalDate fechaFinal;
    private Divisa divisa;
    private boolean cumplida;
    private double interesesPendientes;
    private LocalDate ultimaFechaPago;
    protected double TEA;
    protected final double baseTEAAlto = 34.49;  //bases para calcular los intereses dependiendo de las clases hijas
    protected final double baseTEABajo = 30.69;

    public Prestamo(Usuario usuario, double valorInicial, int tiempo, LocalDate fechaInicio, LocalDate fechaFinal, Divisa divisa) {
        setUsuario(usuario);
        setValorInicial(valorInicial);
        valorPagado = 0;
        setDivisa(divisa);
        setTiempo(tiempo);
        this.fechaInicio = fechaInicio;
        setUltimaFechaPago(fechaInicio);
        setFechaPago(fechaFinal);
        setCumplida(false);
        setInteresesPendientes(0);
    }

    public Usuario getUsuario() {
        return usuario;
    }

    public void setUsuario(Usuario usuario) {
        this.usuario = usuario;
    }

    public LocalDate getFechaFinal() {
        return fechaFinal;
    }

    public void setFechaFinal(LocalDate fechaFinal) {
        this.fechaFinal = fechaFinal;
    }

    public LocalDate getUltimaFechaPago() {
        return ultimaFechaPago;
    }

    public void setUltimaFechaPago(LocalDate ultimaFechaPago) {
        this.ultimaFechaPago = ultimaFechaPago;
    }

    public double getTEA() {
        return TEA;
    }

    protected abstract void setTEA(double monto);

    protected abstract void setTEA(double monto, Garantia garantia);

    public double getValorInicial() {
        return valorInicial;
    }

    public void setValorInicial(double valorInicial) {
        this.valorInicial = valorInicial;
    }

    public double getValorPagado() {
        return valorPagado;
    }

    public void setValorPagado(double valorPagado) {
        this.valorPagado = valorPagado;
    }

    public int getTiempo() {
        return tiempo;
    }

    public void setTiempo(int tiempo) {
        this.tiempo = tiempo;
    }

    public LocalDate getFechaInicio() {
        return fechaInicio;
    }

    public LocalDate getFechaPago() {
        return fechaFinal;
    }

    public void setFechaPago(LocalDate fechaPago) {
        this.fechaFinal = fechaPago;
    }

    public Divisa getDivisa() {
        return divisa;
    }

    public void setDivisa(Divisa divisa) {
        this.divisa = divisa;
    }

    public boolean isCumplida() {
        return cumplida;
    }

    public double getInteresesPendientes() {
        return interesesPendientes;
    }

    public void setInteresesPendientes(double interesesPendientes) {
        this.interesesPendientes = interesesPendientes;
    }

    public void setCumplida(boolean cumplida) {
        this.cumplida = cumplida;
    }

    public double calcularCuotas() {
        long n = Utils.mesesEntreFechas(this.ultimaFechaPago, this.fechaFinal);
        double capitalPendiente = this.interesesPendientes + this.valorInicial - this.valorPagado;
        double r = Math.pow(1 + this.TEA / 100, 1 / 12.0) - 1;
        double cuotaMensual = (r * capitalPendiente) / (1 - Math.pow(1 + r, -n));
        return cuotaMensual;
    }

    public double[] abonar(double monto, Cuenta origen) {
    	double[] arreglo = new double[3];
        if (!cumplida) {
            double capitalPendiente = this.valorInicial - this.valorPagado;
            long diasDeIntereses = Utils.diasEntreFechas(this.ultimaFechaPago, LocalDate.now());
            double intereses = interesesPendientes + ((this.interesesPendientes + capitalPendiente) * (Math.pow(1 + (this.TEA / 100), diasDeIntereses / 365.0) - 1));
            double pendiente = intereses + capitalPendiente;
            LocalDate now = LocalDate.now();
            double[] monto2 = origen.getDivisa().ConvertToDivisa(monto, this.divisa);
            Salida salida = new Salida(monto2[0],monto, now, origen, null, origen.getDivisa(), this.divisa);
            if (this.usuario.nuevaSalida(salida)) {
                double nuevosIntereses = Math.max(0, intereses - monto2[0]);
                double nuevoCapital = Math.min(capitalPendiente, pendiente - monto2[0]);
                arreglo[0] = intereses - nuevosIntereses;
                arreglo[1] = Math.max(0, monto2[0] - intereses + nuevosIntereses);
                arreglo[2] = monto2[1];
                this.interesesPendientes = nuevosIntereses;
                this.valorPagado = this.valorInicial - nuevoCapital;
                this.ultimaFechaPago = LocalDate.now();
                return arreglo;                
            }
            return null;
        }
        return null;
    }

    public Movimiento terminar(Cuenta cuenta) {
    	 this.cumplida = true;
         if (valorPagado > valorInicial) {
             Ingreso ingreso = new Ingreso(this.divisa.ConvertToDivisa(valorPagado - valorInicial, cuenta.getDivisa())[0],(valorPagado-valorInicial), LocalDate.now(), true, null, null, cuenta, this.divisa, cuenta.getDivisa());
             usuario.nuevoIngreso(ingreso);
             return ingreso;
         }
         return null;
    }
}

File: Python/gestorAplicacion/economia/Prestamo.py
from administrador import Utils
from gestorAplicacion.economia import Abonable
from gestorAplicacion.economia import Ingreso
from gestorAplicacion.economia import Salida
from datetime import date, time, datetime

class Prestamo(Abonable):

    def __init__(self, usuario, valorInicial, tiempo, fechaInicio, fechaFinal, divisa):
        self._usuario = usuario
        self._valorInicial = valorInicial
        self._valorPagado = 0
        self._divisa = divisa
        self._tiempo = tiempo
        self._fechaInicio = fechaInicio
        self._ultimaFechaPago = fechaInicio
        self._fechaPago = fechaFinal
        self._cumplida = False
        self._interesesPendientes = 0
        self._ultimaFechaPago = None
        self._TEA = 0
        self._baseTEAAlto = 34.49
        self._baseTEABajo = 30.69


    def getUsuario(self):
        return self._usuario

    def setUsuario(self, usuario):
        self._usuario = usuario

    def getFechaFinal(self):
        return self._fechaFinal

    def setFechaFinal(self, fechaFinal):
        self._fechaFinal = fechaFinal

    def getUltimaFechaPago(self):
        return self._ultimaFechaPago

    def setUltimaFechaPago(self, ultimaFechaPago):
        self._ultimaFechaPago = ultimaFechaPago

    def getTEA(self):
        return self._TEA

    def getValorInicial(self):
        return self._valorInicial

    def setValorInicial(self, valorInicial):
        self._valorInicial = valorInicial

    def getValorPagado(self):
        return self._valorPagado

    def setValorPagado(self, valorPagado):
        self._valorPagado = valorPagado

    def getTiempo(self):
        return self._tiempo

    def setTiempo(self, tiempo):
        self._tiempo = tiempo

    def getFechaInicio(self):
        return self._fechaInicio

    def getFechaPago(self):
        return self._fechaFinal

    def setFechaPago(self, fechaPago):
        self._fechaFinal = fechaPago

    def getDivisa(self):
        return self._divisa

    def setDivisa(self, divisa):
        self._divisa = divisa

    def isCumplida(self):
        return self._cumplida

    def getInteresesPendientes(self):
        return self._interesesPendientes

    def setInteresesPendientes(self, interesesPendientes):
        self._interesesPendientes = interesesPendientes

    def setCumplida(self, cumplida):
        self._cumplida = cumplida

    def calcularCuotas(self):
        n = Utils.mesesEntreFechas(self.ultimaFechaPago, self.fechaFinal)
        capitalPendiente = self._interesesPendientes + self._valorInicial - self._valorPagado
        r = (1 + self.TEA / 100) ** (1 / 12.0) - 1
        cuotaMensual = (r * capitalPendiente) / (1 - (1 + r) ** (-n))
        return cuotaMensual

    def abonar(self, monto, origen):
        arreglo = [0 for _ in range(3)]
        if self._cumplida == False:
            capitalPendiente = self.valorInicial - self.valorPagado
            diasDeIntereses = Utils.diasEntreFechas(self.ultimaFechaPago, date.today())
            intereses = self._interesesPendientes + ((self._interesesPendientes + capitalPendiente) * ((1 + (self.TEA / 100)) ** (diasDeIntereses / 365.0) - 1))
            pendiente = intereses + capitalPendiente
            now = date.today()
            monto2 = origen.getDivisa().ConvertToDivisa(monto, self.divisa)
            salida = Salida(monto2[0],monto, now, origen, None, origen.getDivisa(), self.divisa)
            if self.usuario.nuevaSalida(salida):
                nuevosIntereses = max(0, intereses - monto2[0])
                nuevoCapital = min(capitalPendiente, pendiente - monto2[0])
                arreglo[0] = intereses - nuevosIntereses
                arreglo[1] = max(0, monto2[0] - intereses + nuevosIntereses)
                arreglo[2] = monto2[1]
                self.interesesPendientes = nuevosIntereses
                self.valorPagado = self.valorInicial - nuevoCapital
                self.ultimaFechaPago = date.today()
                return arreglo
            return None
        return None

    def terminar(self, cuenta):
        self._cumplida = True
        if self._valorPagado > self._valorInicial:
            ingreso = Ingreso(self.divisa.ConvertToDivisa(self._valorPagado - self._valorInicial, cuenta.getDivisa())[0],(self._valorPagado-self._valorInicial), date.today(), True, None, None, cuenta, self.divisa, cuenta.getDivisa())
            self._usuario.nuevoIngreso(ingreso)
            return ingreso
        return None
